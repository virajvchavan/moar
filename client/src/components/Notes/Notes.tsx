import { createStyles, makeStyles, Theme } from "@material-ui/core";
import React from "react";
import ContentEditable from "react-contenteditable";
import sanitizeHtml from "sanitize-html";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    editable: {
      fontFamily: 'sans-serif',
      whiteSpace: 'pre-wrap',
      wordBreak: 'keep-all',
      width: "100%",
      border: 'none',
      padding: '5px',
      paddingBottom: 0,
      resize: 'none',
      margin: 0,
      outline: '0px solid transparent'
    },
    child_notes: {
      marginLeft: "20px"
    },
    noteRow: {
      display: "inline-flex",
      width: "100%",
      '&:hover': {
        '& $expander': {
          color: 'black'
        }
      }
    },
    bullet: {
      width: "18px",
      height: "18px",
      lineHeight: '37px',
      cursor: "pointer"
    },
    expander: {
      width: "15px",
      height: "20px",
      color: '#cecece',
      transition: 'transform 200ms ease 0s',
      lineHeight: '40px'
    },
    expanded: {
      transform: 'rotateZ(90deg)'
    },
    emptySpace: {
      width: "15px"
    }
  }),
);

export interface NotesType {
  id: string;
  content: string;
  child_notes: Array<NotesType>,
  collapsed? : boolean
}

interface Props {
  notes: Array<NotesType>
  className?: string,
  index: string,
  onNoteContentChange: (deepIndex: string, newContent: string) => void,
  addAChildNote: (deepIndex: string) => void,
  handleTabPress: (deepIndex: string) => void,
  handleBackspaceWhenEmpty: (evt: React.KeyboardEvent<HTMLDivElement>, deepIndex: string) => void,
  handleUpKey: (deepIndex: string) => void
  handleDownKey: (deepIndex: string) => void,
  setCollapsedForNote: (deepIndex: string, state: boolean) => void,
  handleShiftTabPress: (deepIndex: string) => void
}

const sanitizeConf = {
  allowedTags: ["b", "i", "em", "strong", "a", "p", "h1"],
  allowedAttributes: { a: ["href"] }
};

// This component is supposed to be a dumb one, all logic related to notes should be handled in RootNote.tsx.
export default function Notes(props: Props) {
  const classes = useStyles();

  return <>
    {props.notes.map((note, index) => {
      let deepIndex = `${props.index}.${index}`;

      const onKeyDown = (evt: React.KeyboardEvent<HTMLDivElement>) => {
        if (evt.key === "Enter" && !evt.shiftKey) {
          evt.preventDefault();
          props.addAChildNote(deepIndex);
        } else if (evt.key === "Tab") {
          evt.preventDefault();
          evt.stopPropagation();
          if (evt.shiftKey) {
            props.handleShiftTabPress(deepIndex);
          } else {
            props.handleTabPress(deepIndex);
          }
        } else if (evt.key === "Backspace") {
          props.handleBackspaceWhenEmpty(evt, deepIndex);
        } else if (evt.key === "ArrowUp") {
          evt.preventDefault();
          props.handleUpKey(deepIndex);
        } else if (evt.key === "ArrowDown") {
          evt.preventDefault();
          props.handleDownKey(deepIndex);
        }
      }

      return <div className={props.className} key={deepIndex}>
        <div className={classes.noteRow}>
          {note.child_notes.length > 0 ? (
            <div className={classes.expander} onClick={() => props.setCollapsedForNote(deepIndex, !note.collapsed)}>
              <svg className={!note.collapsed ? classes.expanded : undefined} width="20" height="20" viewBox="0 0 20 20" ><path d="M13.75 9.56879C14.0833 9.76124 14.0833 10.2424 13.75 10.4348L8.5 13.4659C8.16667 13.6584 7.75 13.4178 7.75 13.0329L7.75 6.97072C7.75 6.58582 8.16667 6.34525 8.5 6.5377L13.75 9.56879Z" stroke="none" fill="currentColor"></path></svg>
            </div>
          ): <div className={classes.emptySpace}></div>}
          <div className={classes.bullet}>
            <svg viewBox="0 0 18 18" fill="#747474"><circle cx="9" cy="9" r="3.5"></circle></svg>
          </div>
          <ContentEditable
            id={'note' + deepIndex}
            className={classes.editable}
            tagName="pre"
            html={note.content} // innerHTML of the editable div
            disabled={false} // use true to disable edition
            onChange={(evt) => props.onNoteContentChange(deepIndex, sanitizeHtml(evt.target.value, sanitizeConf))} // handle innerHTML change
            onBlur={(evt) => console.log("blurred")}
            onKeyDown={onKeyDown}
          />
        </div>

        {!note.collapsed && note.child_notes ?
          <Notes
            index={deepIndex}
            className={classes.child_notes}
            notes={note.child_notes}
            onNoteContentChange={props.onNoteContentChange}
            addAChildNote={props.addAChildNote}
            handleTabPress={props.handleTabPress}
            handleBackspaceWhenEmpty={props.handleBackspaceWhenEmpty}
            handleUpKey={props.handleUpKey}
            handleDownKey={props.handleDownKey}
            setCollapsedForNote={props.setCollapsedForNote}
            handleShiftTabPress={props.handleShiftTabPress}
          />
          : null
        }
      </div>
    })}
  </>
}
