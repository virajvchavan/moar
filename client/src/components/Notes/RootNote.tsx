import React, { useEffect, useState } from "react";
import { makeStyles, Theme, createStyles, Paper } from "@material-ui/core";
import Notes, { NotesType } from "./Notes";
import { useAuth } from '../../hooks/use-auth';
import produce from 'immer';
import { useDebounce } from 'use-debounce';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    notesRoot: {
      marginLeft: "40px",
      paddingTop: "10px",
      paddingBottom: "10px"
    },
    paper: {
      width: "100%"
    },
    addBtn: {
      marginLeft: "20px",
      cursor: "pointer",
      color: 'grey',
      marginTop: "5px"
    }
  }),
);

export default function RootNotes() {
  const classes = useStyles();
  const auth = useAuth();
  const [notes, setNotes] = useState<Array<NotesType>>([]);
  const [syncedNotes, setSyncedNotes] = useState<Array<NotesType>>([]);

  // if the user doesn't type anything for 5 seconds straight, make API calls to the server
  const [debouncedNotes] = useDebounce(notes, 5000);

  useEffect(() => {
    if (syncedNotes.length > 0 && debouncedNotes.length > 0) {
      // console.log(syncedNotes);
      // console.log(debouncedNotes);
      console.log("calling the api");
    }
  }, [debouncedNotes, syncedNotes])

  useEffect(() => {
    window.fetch('/api/notes', {
      method: 'GET',
      credentials: 'include',
      headers: {
          'Authorization': "Bearer " + auth?.user?.token,
          'Content-Type': 'application/json'
      }
    })
    .then(response => response.json())
    .then(json => {
      setNotes(json);
      setSyncedNotes(json);
    })
    .catch(error => console.log(error));
  }, [auth]);

  const updateNoteField = (deepIndex: string, field: 'content' | 'collapsed', valueToSet: any) => {
    setNotes(produce(newNotes => {
      let indices = deepIndex.slice(1).split(".").map(i => parseInt(i));
      let noteToUpdate = newNotes[indices[0]];
      indices.shift();
      indices.forEach(index => {
        noteToUpdate = noteToUpdate.child_notes[index];
      });
      if (field === "content") {
        noteToUpdate.content = valueToSet;
      } else if (field === "collapsed") {
        noteToUpdate.collapsed = valueToSet;
      }
    }));
  }

  const onNoteContentChange = (deepIndex: string, newContent: string) => {
    updateNoteField(deepIndex, "content", newContent);
  }

  const setCollapsedForNote = (deepIndex: string, state: boolean) => {
    updateNoteField(deepIndex, "collapsed", state);
    focusOnANote(deepIndex);
  }

  // indices are the sequence in which to access a note in the state
  const getNoteForIndices = (newNotes: NotesType[], indices: number[]) => {
    let noteToUpdate = newNotes[indices[0]];
    indices.forEach((index, i) => {
      if (i !== 0) {
        noteToUpdate = noteToUpdate.child_notes[index];
      }
    });
    return noteToUpdate;
  }

  const addAChildNote = (deepIndex: string) => {
    setNotes(produce(newNotes => {
      let indices = deepIndex.slice(1).split(".").map(i => parseInt(i));
      let emptyNote = {content: "", id: "", child_notes: []};
      let noteIndexToFocusOn = "";

      let currentNote = getNoteForIndices(newNotes, indices);
      let childrenOfCurrentNote = currentNote.child_notes;
      if (childrenOfCurrentNote.length > 0 && !currentNote.collapsed) {
        childrenOfCurrentNote.unshift(emptyNote);
        noteIndexToFocusOn = deepIndex + ".0";
      } else {
        if (indices.length > 1) {
          let newNoteIndex: number = indices.pop() || 0;
          newNoteIndex += 1;
          let noteToUpdate = getNoteForIndices(newNotes, indices);
          noteToUpdate.child_notes.splice(newNoteIndex || 0, 0, emptyNote);
          noteToUpdate.collapsed = false;
          noteIndexToFocusOn = `.${indices.join(".")}.${newNoteIndex}`;
        } else {
          newNotes.splice(indices[0] + 1, 0, emptyNote);
          noteIndexToFocusOn = `.${indices[0] + 1}`;
        }
      }
      focusOnANote(noteIndexToFocusOn);
    }));
  }

  const handleTabPress = (deepIndex: string) => {
    setNotes(produce(newNotes => {
      let indices = deepIndex.slice(1).split(".").map(i => parseInt(i));
      let originalIndex = indices.pop() || 0;
      let newLeafIndex: number;
      if (originalIndex > 0) {
        if (indices.length === 0) {
          let noteToMove = newNotes.splice(originalIndex, 1)[0];
          newNotes[originalIndex - 1].child_notes.push(noteToMove);
          newNotes[originalIndex - 1].collapsed = false;
          newLeafIndex = newNotes[originalIndex - 1].child_notes.length - 1;
        } else {
          let originalParent = getNoteForIndices(newNotes, indices);
          let noteToMove = originalParent.child_notes.splice(originalIndex, 1)[0];
          originalParent.child_notes[originalIndex - 1].child_notes.push(noteToMove);
          originalParent.child_notes[originalIndex - 1].collapsed = false;
          newLeafIndex = originalParent.child_notes[originalIndex - 1].child_notes.length - 1;
        }
        let indexToFocuson = indices.length > 0 ? `.${indices.join(".")}` : "";
        indexToFocuson += `.${originalIndex - 1}.${newLeafIndex}`;
        focusOnANote(indexToFocuson);
      }
    }));
  }

  const handleShiftTabPress = (deepIndex: string) => {
    setNotes(produce(newNotes => {
      let indices = deepIndex.slice(1).split(".").map(i => parseInt(i));

      if (indices.length === 1) return;

      let originalIndex = indices.pop() || 0;
      let parentIndex = indices.pop() || 0;
      if (indices.length === 0) {
        newNotes.splice(parentIndex + 1, 0, newNotes[parentIndex].child_notes.splice(originalIndex, 1)[0]);
      } else {
        let grandParentNote = getNoteForIndices(newNotes, indices);
        grandParentNote.child_notes.splice(
          parentIndex + 1,
          0,
          grandParentNote.child_notes[parentIndex].child_notes.splice(originalIndex, 1)[0]
        )
      }
      let indexToFocusOn = indices.length > 0 ? `.${indices.join(".")}` : "";
      indexToFocusOn += `.${parentIndex + 1}`;
      focusOnANote(indexToFocusOn);
    }));
  }

  const handleBackspaceWhenEmpty = (deepIndex: string) => {
    setNotes(produce((newNotes: NotesType[]) => {
      let indices = deepIndex.slice(1).split(".").map(i => parseInt(i));
      let originalIndex = indices.pop() || 0;
      if (indices.length === 0) {
        let currentNote = newNotes[originalIndex];
        if (!currentNote.content) {
          if (currentNote.child_notes.length > 0) {
            newNotes.push(...currentNote.child_notes);
          }
          newNotes.splice(originalIndex, 1);
          focusOnANote(`.${originalIndex - 1}`);
        }
      } else {
        let parentNote = getNoteForIndices(newNotes, indices);
        let currentNote = parentNote.child_notes[originalIndex];
        if (!currentNote.content) {
          if (currentNote.child_notes.length > 0) {
            // add its children to its parent
            parentNote.child_notes.push(...currentNote.child_notes);
          }
          parentNote.child_notes.splice(originalIndex, 1);

          let indexToFocusOn = `.${indices.join(".")}`;
          if (originalIndex > 0) {
            indexToFocusOn += `.${originalIndex - 1}`;
          }
          focusOnANote(indexToFocusOn);
        }
      }
    }));
  }

  const handleUpKey = (deepIndex: string) => {
    setNotes(produce(newNotes => {
      let indices = deepIndex.slice(1).split(".").map(i => parseInt(i));
      if (indices[indices.length - 1] > 0) {
        // if has any siblings
        indices[indices.length - 1] = indices[indices.length - 1] -1;
        let noteToFocus = getNoteForIndices(newNotes, indices);
        while (noteToFocus.child_notes.length > 0 && !noteToFocus.collapsed) {
          indices.push(noteToFocus.child_notes.length - 1);
          noteToFocus = noteToFocus.child_notes[noteToFocus.child_notes.length - 1]
        }
      } else {
        indices.pop();
      }
      focusOnANote(`.${indices.join(".")}`);
    }));
  }

  const handleDownKey = (deepIndex: string) => {
    setNotes(produce(newNotes => {
      let indices = deepIndex.slice(1).split(".").map(i => parseInt(i));
      let originalIndex = indices.pop() || 0;
      if (indices.length === 0) {
        let currentNote = newNotes[originalIndex];
        if (currentNote.child_notes.length > 0 && !currentNote.collapsed) {
          indices.push(originalIndex);
          indices.push(0);
        } else {
          indices.push(originalIndex + 1);
        }
      } else {
        let parentNote = getNoteForIndices(newNotes, indices);
        let currentNote = parentNote.child_notes[originalIndex];
        if (currentNote.child_notes.length > 0 && !currentNote.collapsed) {
          indices.push(originalIndex);
          indices.push(0);
        } else {
          indices.push(originalIndex);
            while (true) {
              indices[indices.length - 1] = indices[indices.length - 1] + 1;
              if (indices.length === 1) {
                break;
              }
              // doing it this way because accessing parent notes of a note would be expensive
              if (document.getElementById(`note.${indices.join(".")}`)) {
                break;
              } else {
                indices.pop();
              }
            }
        }
      }
      focusOnANote(`.${indices.join(".")}`);
    }))
  }

  const onAddBtnClick = () => {
    setNotes(produce(newNotes => {
      newNotes.push({content: "", id: "", child_notes: []});
      focusOnANote(`.${newNotes.length - 1}`);
    }));
  }

  // directly accessing dom here to avoid passing refs in an infinitely nested list
  // passing refs too deep might be a performance issue as well
  const focusOnANote = (deepIndex: string) => {
    setTimeout(() => {
      document.getElementById("note" + deepIndex)?.focus();
    }, 0);
  }

  return <Paper elevation={2} className={classes.paper}>
    <div className={classes.notesRoot}>
      <Notes notes={notes} index=""
        onNoteContentChange={onNoteContentChange}
        addAChildNote={addAChildNote}
        handleTabPress={handleTabPress}
        handleBackspaceWhenEmpty={handleBackspaceWhenEmpty}
        handleUpKey={handleUpKey}
        handleDownKey={handleDownKey}
        setCollapsedForNote={setCollapsedForNote}
        handleShiftTabPress={handleShiftTabPress}
      />
      <div className={classes.addBtn} onClick={onAddBtnClick}>+</div>
    </div>
  </Paper>
}
