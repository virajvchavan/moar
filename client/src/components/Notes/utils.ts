import { NotesType } from './Notes';

let jsonDiff = require('jsondiffpatch').create({
  objectHash: (obj: NotesType) => obj.id,
  textDiff: {
    minLength: 1000000
  }
});

interface SyncedDataResponse {
  indexPath: string[],
  newId: string
}

type NoteFields = {
  content? : string,
  collapsed?: string
}

interface Transaction {
  type: "add" | "delete" | "move_same_parent" | "update",
  id?: string,
  new_index?: number,
  parent_id?: string,
  fields?: NoteFields
}

  // indices are the sequence in which to access a note in the state
  const getNoteForIndices = (newNotes: NotesType[], indices: number[]) => {
    let noteToUpdate = newNotes[indices[0]];
    indices.forEach((index, i) => {
      if (i !== 0) {
        noteToUpdate = noteToUpdate.child_notes[index];
      }
    });
    return noteToUpdate;
  }

const generateTransactions = (key: string, changes: JSON, indexes: Array<number>, newNotes: NotesType[]): Transaction[] => {
  let transactions: Transaction[] = [];
  if (key[0] === "_") {
    // it's either a delete or move
    let indexFromLhs = key.slice(1);
    if (Array.isArray(changes)) {
      if (changes.length === 2) {
        let noteId = getNoteForIndices(newNotes, indexes).child_notes[changes[1]].id;
        transactions.push({ type: "move_same_parent", id: noteId, new_index: changes[1] });
      } else if (changes.length === 3) {
        transactions.push({ type: "delete", id: changes[0].id });
      }
    }
  } else {
    // it's either an update for existing note or it's an addition of a note
    if (Array.isArray(changes)) {
      let parent_id: string = "root";
      if (indexes.length > 0) {
        console.log("getting parent from: " + indexes);
        parent_id = getNoteForIndices(newNotes, indexes).id;
      }
      transactions.push({type: "add", parent_id: parent_id, id: changes[0].id, new_index: parseInt(key), fields: changes[0]});
    } else {
      // update the content/collapsed
      if ("child_notes" in changes) {
        transactions.push(...createTransactionsFromChanges(changes["child_notes"], [...indexes, parseInt(key)], newNotes));
      }

      let fields: NoteFields = {};
      if ("content" in changes) {
        fields.content = changes["content"][1] ? changes["content"][1] : changes["content"][0];
      }
      if ("collapsed" in changes) {
        fields.collapsed = changes["collapsed"][1] ? changes["collapsed"][1] : changes["collapsed"][0];
      }
      if (Object.keys(fields).length > 0) {
        transactions.push({type: "update", id: "asdad", fields: fields});
      }
    }
  }
  return transactions;
}

export const syncChangesWithServer = async (newNotes: NotesType[], syncedNotes: NotesType[]) => {
  console.log("calling the api");
  let changes = jsonDiff.diff(syncedNotes, newNotes);

  if (!changes) return false;

  console.log(createTransactionsFromChanges(changes, [], newNotes));
  return true;
}

function createTransactionsFromChanges(changes: any, indexes: Array<number>, newNotes: NotesType[]) {
  let transactions: Transaction[] = [];
  if (changes["_t"] && changes["_t"] === "a") {
    changes && Object.keys(changes).forEach(key => {
      if (key !== "_t") {
        transactions.push(...generateTransactions(key, changes[key], indexes, newNotes));
      }
    });
  }
  return transactions;
}
