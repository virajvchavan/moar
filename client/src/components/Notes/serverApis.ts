import { Transactions } from './utils';

export const fetchAllNotes = async (auth_token: string) => {
  let response = await window.fetch('/api/notes', {
    method: 'GET',
    credentials: 'include',
    headers: {
      'Authorization': "Bearer " + auth_token,
      'Content-Type': 'application/json'
    }
  });
  return await response.json();
};

export async function callProcessTransactionsApi(authToken: string, transactions: Transactions) {
  return await fetch("/api/notes/process_transactions", {
    method: "POST", credentials: 'include',
    headers: {
      'Authorization': "Bearer " + authToken,
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(transactions)
  });
}
