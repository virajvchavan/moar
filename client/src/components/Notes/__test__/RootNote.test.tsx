import { cleanup, render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from "react";
import RootNotes from '../RootNote';
import * as utils from '../utils';
import * as serverApis from '../serverApis';
import { NotesType } from '../Notes';

afterEach(cleanup);

// Note: writing tests for any components that use content-editable is not well supported right now
// See: https://github.com/testing-library/user-event/issues/230
// Some possible test cases that are missing because of this issue are: 
// - keyboard navigation testing, - anything that requires typing in content-editable component

let someNotes = [
  {"content":"one","child_notes":[],"collapsed":false,"id":"1"},
  {"content":"two","child_notes":[],"collapsed":false,"id":"2"},
  {"content":"three","child_notes":[],"collapsed":false,"id":"3"}
];

function mockApiCalls(notes: NotesType[]) {
  jest.spyOn(serverApis, 'fetchAllNotes').mockImplementation(async () => notes);
  jest.spyOn(serverApis, 'callProcessTransactionsApi').mockImplementation(async () => new Response("[]", {status: 200, headers: {responseType: "application/json"}}));
}

jest.mock('../../../hooks/use-auth.tsx', () => ({
  useAuth: () => {
    return {
      loading: false,
      user: { name: "TestUser", token: "fakeAuthToken" },
      signin: jest.fn(),
      signout: jest.fn(),
      signup: jest.fn()
    }
  }
}));

describe('Initial rendering', () => {
  test("it should display the loading text before making the api call", async () => {
    mockApiCalls(someNotes);
    const { findByText } = render(<RootNotes />);
    let res = await findByText("Loading...");
    expect(res).not.toBeNull();
  });

  test("it should display the Add(+) btn", async () => {
    mockApiCalls(someNotes);
    const { findByText } = render(<RootNotes />);
    let res = await findByText("+");
    expect(res).not.toBeNull();
  });

  test("it should render initial notes correctly when max level is 1", async () => {
    mockApiCalls(someNotes);
    const { findByText } = render(<RootNotes />);
    let note1 = await findByText("one");
    expect(note1.id).toBe("note.0");

    let note2 = await findByText("two");
    expect(note2.id).toBe("note.1");

    let note3 = await findByText("three");
    expect(note3.id).toBe("note.2");
  });

  test("it should render initial notes correctly when max level is more than 1", async () => {
    let notes = [
      {"content":"one","child_notes":[],"collapsed":false,"id":"1"},
      {"content":"two","child_notes":[{"content":"three","child_notes":[
        {"content":"four","child_notes":[],"collapsed":false,"id":"4"}
      ],"collapsed":false,"id":"3"}],"collapsed":false,"id":"2"},
    ]
    mockApiCalls(notes);

    const { findByText } = render(<RootNotes />);
    let note1 = await findByText("one");
    expect(note1.id).toBe("note.0");

    let note2 = await findByText("two");
    expect(note2.id).toBe("note.1");

    let note3 = await findByText("three");
    expect(note3.id).toBe("note.1.0");

    let note4 = await findByText("four");
    expect(note4.id).toBe("note.1.0.0");
  });
});

describe('Collapsing logic', () => {
  test("it should not render children of a collapsed note", async () => {
    let notes = [
      {"content":"two","child_notes":[{"content":"three","child_notes": [],"collapsed":false,"id":"3"}],"collapsed":true,"id":"2"},
    ]
    mockApiCalls(notes);

    const { queryByText, findByText } = render(<RootNotes />);

    await waitFor(() => findByText("two"));
    expect(queryByText("three")).toBeNull();
  });

  // TODO:
  // test("it should collapse child notes after clicking on the collapse arrow", async () => {
  // });
});

describe('Adding new notes', () => {
  test("it should add a new note at root level when + btn is clicked", async () => {
    let notes = [
      {"content":"two","child_notes":[{"content":"three","child_notes": [],"collapsed":false,"id":"3"}],"collapsed":false,"id":"2"},
    ]
    mockApiCalls(notes);

    const { findByTestId, findByText } = render(<RootNotes />);
    let addBtn = await findByText("+");
    userEvent.click(addBtn);
    let note = await findByTestId("noterow.1");
    expect(note).not.toBeNull();
  });

  test("it should call syncChangesWithServer after 4 seconds of updating notes", async () => {
    let notes = [
      {"content":"two","child_notes":[{"content":"three","child_notes": [],"collapsed":false,"id":"3"}],"collapsed":false,"id":"2"},
    ]
    mockApiCalls(notes);

    const { findByText, findByTestId } = render(<RootNotes />);
    let addBtn = await findByText("+");

    userEvent.click(addBtn);
    await findByTestId("noterow.1");
    setTimeout(() => {
      expect(utils.syncChangesWithServer).toHaveBeenCalled();
    }, 4000);
  });

  test("it should add a new note at root level when enter key is pressed on a root note with no child notes", async () => {
    let notes = [
      {"content":"one","child_notes":[],"collapsed":false,"id":"1"},
    ]
    mockApiCalls(notes);

    const { findByTestId, findByText } = render(<RootNotes />);
    let note1 = await waitFor(() => findByText("one"));
    userEvent.type(note1, '{enter}');
    let note = await findByTestId("noterow.1");
    expect(note).not.toBeNull();
  });

  test("it should add a new note at root level when enter key is pressed on a root note with collapsed child notes", async () => {
    let notes = [
      {"content":"one","child_notes":[{"content":"two","child_notes": [],"collapsed":false,"id":"2"}],"collapsed":true,"id":"1"},
    ]
    mockApiCalls(notes);

    const { findByTestId, findByText } = render(<RootNotes />);
    let note1 = await waitFor(() => findByText("one"));
    userEvent.type(note1, '{enter}');
    let note = await findByTestId("noterow.1");
    expect(note).not.toBeNull();
  });

  test("it should add a new note as the first child when enter key is pressed on a root note with non-collapsed child notes", async () => {
    let notes = [
      {"content":"one","child_notes":[{"content":"two","child_notes": [],"collapsed":false,"id":"2"}],"collapsed":false,"id":"1"},
    ]
    mockApiCalls(notes);

    const { findByTestId, findByText } = render(<RootNotes />);
    let note1 = await waitFor(() => findByText("one"));
    userEvent.type(note1, '{enter}');
    let note = await findByTestId("noterow.0.1");
    expect(note).toHaveTextContent("two");
  });


  test("it should add a new note at the correct index when enter key pressed in some middle note", async () => {
    let notes = [
      {"content":"one","child_notes":[],"collapsed":false,"id":"1"},
      {"content":"two","child_notes":[],"collapsed":false,"id":"2"},
    ]
    mockApiCalls(notes);

    const { findByTestId, findByText } = render(<RootNotes />);
    let note1 = await waitFor(() => findByText("one"));
    userEvent.type(note1, '{enter}');

    let note = await findByTestId("noterow.2");
    expect(note.querySelector("pre")).toHaveTextContent("two");
  });
});

describe('Tab key-press logic', () => {
  test("it should change the parent of a note if tab is pressed and the note has a sibling before it", async () => {
    let notes = [
      {"content":"one","child_notes":[],"collapsed":false,"id":"1"},
      {"content":"two","child_notes":[],"collapsed":false,"id":"2"},
    ]
    mockApiCalls(notes);

    const { findByTestId, findByText } = render(<RootNotes />);
    let note = await waitFor(() => findByText("two"));
    note.focus();
    expect(note).toHaveFocus();
    userEvent.tab();

    let newNote = await findByTestId("noterow.0.0");
    expect(newNote.querySelector("pre")).toHaveTextContent("two");
  });

  test("it should change the parent of a note if shift+tab is pressed and the note has a parent", async () => {
    let notes = [
      {"content":"one","child_notes":[{"content":"two","child_notes":[],"collapsed":false,"id":"2"}],"collapsed":false,"id":"1"},
    ]
    mockApiCalls(notes);

    const { findByTestId, findByText } = render(<RootNotes />);
    let note = await waitFor(() => findByText("two"));
    note.focus();
    expect(note).toHaveFocus();
    userEvent.tab({ shift : true });

    let newNote = await findByTestId("noterow.1");
    expect(newNote.querySelector("pre")).toHaveTextContent("two");
  });
});

// TODO
// stuck bcause triggering the 'up' & 'down' arrows by using 'user-event' is not supported for content-editable elements right now
// https://github.com/testing-library/user-event/issues/230
describe('Navigating between notes using up and down arrows', () => {
  // test("it should change the focus to the correct note when down arrow key is pressed when note has no children", async () => {
  // });

  // test("it should change the focus to the correct note when down arrow key is pressed when note has collapsed children", async () => {
  // });

  // test("it should change the focus to the correct note when up arrow key is pressed when note has siblings before it", async () => {
  // });

  // test("it should change the focus to the correct note when up arrow key is pressed when it's the first child", async () => {
  // });
});

// TODO
describe('Deleting notes', () => {
  // test("it should delete the note when backspace is pressed when the note is empty", async () => {
  // });

  // test("it should assign the children of a deleted note to the previous sibling of it ", async () => {
  // });
});
