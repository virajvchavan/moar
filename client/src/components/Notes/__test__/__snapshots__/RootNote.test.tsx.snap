// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots it should match snapshot for the initial render with some notes 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root makeStyles-paper-132 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="makeStyles-notesRoot-131"
    >
      <div
        data-testid="noterow.0"
      >
        <div
          class="makeStyles-noteRow-136"
        >
          <div
            class="makeStyles-expander-138"
            data-testid="collapseBtn.0"
          >
            <svg
              class="makeStyles-expanded-139"
              height="20"
              viewBox="0 0 20 20"
              width="20"
            >
              <path
                d="M13.75 9.56879C14.0833 9.76124 14.0833 10.2424 13.75 10.4348L8.5 13.4659C8.16667 13.6584 7.75 13.4178 7.75 13.0329L7.75 6.97072C7.75 6.58582 8.16667 6.34525 8.5 6.5377L13.75 9.56879Z"
                fill="currentColor"
                stroke="none"
              />
            </svg>
          </div>
          <div
            class="makeStyles-bullet-137"
          >
            <svg
              fill="#747474"
              viewBox="0 0 18 18"
            >
              <circle
                cx="9"
                cy="9"
                r="3.5"
              />
            </svg>
          </div>
          <pre
            class="makeStyles-editable-134"
            contenteditable="true"
            id="note.0"
          >
            one
          </pre>
        </div>
        <div
          class="makeStyles-child_notes-135"
          data-testid="noterow.0.0"
        >
          <div
            class="makeStyles-noteRow-136"
          >
            <div
              class="makeStyles-emptySpace-140"
            />
            <div
              class="makeStyles-bullet-137"
            >
              <svg
                fill="#747474"
                viewBox="0 0 18 18"
              >
                <circle
                  cx="9"
                  cy="9"
                  r="3.5"
                />
              </svg>
            </div>
            <pre
              class="makeStyles-editable-134"
              contenteditable="true"
              id="note.0.0"
            >
              one one
            </pre>
          </div>
        </div>
        <div
          class="makeStyles-child_notes-135"
          data-testid="noterow.0.1"
        >
          <div
            class="makeStyles-noteRow-136"
          >
            <div
              class="makeStyles-emptySpace-140"
            />
            <div
              class="makeStyles-bullet-137"
            >
              <svg
                fill="#747474"
                viewBox="0 0 18 18"
              >
                <circle
                  cx="9"
                  cy="9"
                  r="3.5"
                />
              </svg>
            </div>
            <pre
              class="makeStyles-editable-134"
              contenteditable="true"
              id="note.0.1"
            >
              one two
            </pre>
          </div>
        </div>
      </div>
      <div
        data-testid="noterow.1"
      >
        <div
          class="makeStyles-noteRow-136"
        >
          <div
            class="makeStyles-emptySpace-140"
          />
          <div
            class="makeStyles-bullet-137"
          >
            <svg
              fill="#747474"
              viewBox="0 0 18 18"
            >
              <circle
                cx="9"
                cy="9"
                r="3.5"
              />
            </svg>
          </div>
          <pre
            class="makeStyles-editable-134"
            contenteditable="true"
            id="note.1"
          >
            two
          </pre>
        </div>
      </div>
      <div
        data-testid="noterow.2"
      >
        <div
          class="makeStyles-noteRow-136"
        >
          <div
            class="makeStyles-expander-138"
            data-testid="collapseBtn.2"
          >
            <svg
              height="20"
              viewBox="0 0 20 20"
              width="20"
            >
              <path
                d="M13.75 9.56879C14.0833 9.76124 14.0833 10.2424 13.75 10.4348L8.5 13.4659C8.16667 13.6584 7.75 13.4178 7.75 13.0329L7.75 6.97072C7.75 6.58582 8.16667 6.34525 8.5 6.5377L13.75 9.56879Z"
                fill="currentColor"
                stroke="none"
              />
            </svg>
          </div>
          <div
            class="makeStyles-bullet-137"
          >
            <svg
              fill="#747474"
              viewBox="0 0 18 18"
            >
              <circle
                cx="9"
                cy="9"
                r="3.5"
              />
            </svg>
          </div>
          <pre
            class="makeStyles-editable-134"
            contenteditable="true"
            id="note.2"
          >
            three
          </pre>
        </div>
      </div>
      <div
        class="makeStyles-addBtn-133"
      >
        +
      </div>
    </div>
  </div>
</DocumentFragment>
`;
